lib/ocpp/v201/charge_point.cpp

```diff
@@ -2337,6 +2398,54 @@ void ChargePoint::handle_delete_certificate_req(Call<DeleteCertificateRequest> c
     this->send<DeleteCertificateResponse>(call_result);
 }
 
+void ChargePoint::handle_get_log_req(Call<GetLogRequest> call) {
+    const GetLogResponse response = this->callbacks.get_log_request_callback(call.msg);
+
+    ocpp::CallResult<GetLogResponse> call_result(response, call.uniqueId);
+    this->send<GetLogResponse>(call_result);
+}
+
+void ChargePoint::handle_customer_information_req(Call<CustomerInformationRequest> call) {
+    CustomerInformationResponse response;
+    const auto& msg = call.msg;
+    response.status = CustomerInformationStatusEnum::Accepted;
+
+    if (!msg.report and !msg.clear) {
+        EVLOG_warning << "CSMS sent CustomerInformation.req with both report and clear flags being false";
+        response.status = CustomerInformationStatusEnum::Rejected;
+    }
+
+    if (!msg.customerCertificate.has_value() and !msg.idToken.has_value() and !msg.customerIdentifier.has_value()) {
+        EVLOG_warning << "CSMS sent CustomerInformation.req without setting one of customerCertificate, idToken, "
+                         "customerIdentifier fields";
+        response.status = CustomerInformationStatusEnum::Invalid;
+    }
+
+    ocpp::CallResult<CustomerInformationResponse> call_result(response, call.uniqueId);
+    this->send<CustomerInformationResponse>(call_result);
+
+    if (response.status == CustomerInformationStatusEnum::Accepted) {
+        std::string data = "";
+        if (msg.report) {
+            data += this->get_customer_information(msg.customerCertificate, msg.idToken, msg.customerIdentifier);
+        }
+        if (msg.clear) {
+            this->clear_customer_information(msg.customerCertificate, msg.idToken, msg.customerIdentifier);
+        }
+
+        const auto max_customer_information_data_length =
+            this->device_model->get_optional_value<bool>(ControllerComponentVariables::MaxCustomerInformationDataLength)
+                .value_or(DEFAULT_MAX_CUSTOMER_INFORMATION_DATA_LENGTH);
+        if (data.length() > max_customer_information_data_length) {
+            EVLOG_warning << "NotifyCustomerInformation.req data field is too large. Cropping it down to: "
+                          << max_customer_information_data_length << "characters";
+            data = data.substr(0, max_customer_information_data_length);
```

Comment:
This is a bit of a nitpick but doing an erase would be more efficient here. You are now first creating a copy and then removing the original string.

```suggestion
            data.erase(max_customer_information_data_length);
```